src/replit_river/common_session.py

```diff
@@ -29,107 +33,69 @@ async def __call__(
+    ACTIVE = 2
+    CLOSING = 3
+    CLOSED = 4
 
 
-async def setup_heartbeat(
-    session_id: str,
-    heartbeat_ms: float,
-    heartbeats_until_dead: int,
-    get_state: Callable[[], SessionState],
-    get_closing_grace_period: Callable[[], float | None],
-    close_websocket: Callable[[], Awaitable[None]],
-    send_message: SendMessage,
-    increment_and_get_heartbeat_misses: Callable[[], int],
+ConnectingStates = set([SessionState.PENDING, SessionState.CONNECTING])
+TerminalStates = set([SessionState.CLOSING, SessionState.CLOSED])
+
+
+async def buffered_message_sender(
+    connection_condition: asyncio.Condition,
+    message_enqueued: asyncio.Semaphore,
+    get_ws: Callable[[], WebSocketCommonProtocol | ClientConnection | None],
+    websocket_closed_callback: Callable[[], Coroutine[Any, Any, None]],
+    get_next_pending: Callable[[], TransportMessage | None],
+    commit: Callable[[TransportMessage], None],
 ) -> None:
-    logger.debug("Start heartbeat")
     while True:
-        await asyncio.sleep(heartbeat_ms / 1000)
-        state = get_state()
-        if state != SessionState.ACTIVE:
+        await message_enqueued.acquire()
+        while (ws := get_ws()) is None:
+            # Block until we have a handle
             logger.debug(
-                "Session is closed, no need to send heartbeat, state : "
-                "%r close_session_after_this: %r",
-                {state},
-                {get_closing_grace_period()},
+                "buffered_message_sender: Waiting until ws is connected (condition=%r)",
+                connection_condition,
             )
-            # session is closing / closed, no need to send heartbeat anymore
-            return
-        try:
-            await send_message(
-                stream_id="heartbeat",
-                # TODO: make this a message class
-                # https://github.com/replit/river/blob/741b1ea6d7600937ad53564e9cf8cd27a92ec36a/transport/message.ts#L42
-                payload={
-                    "ack": 0,
-                },
-                control_flags=ACK_BIT,
-                procedure_name=None,
-                service_name=None,
-                span=None,
+            async with connection_condition:
+                await connection_condition.wait()
+        if msg := get_next_pending():
+            logger.debug(
+                "buffered_message_sender: Dequeued %r to send over %r",
+                msg,
+                ws,
             )
-
-            if increment_and_get_heartbeat_misses() > heartbeats_until_dead:
-                if get_closing_grace_period() is not None:
-                    # already in grace period, no need to set again
-                    continue
-                logger.info(
-                    "%r closing websocket because of heartbeat misses",
-                    session_id,
+            try:
+                await send_transport_message(msg, ws, websocket_closed_callback)
+                commit(msg)
```

Comment:
what is commit doing

Reply:
Dead code! Deleting!

In the other version of this function, `commit` "accepts" a message as having been sent.